### ISSUE TITLE
[Bug]: toxic `proc-macro2` version

### ISSUE THREAD

**Issue by gutenfries** [corrosion-rs/corrosion#428]
# Description

Hi folks,
As you may or may not be aware, there was recently a push to prod made by the Rust dev team (see https://github.com/rust-lang/rust/issues/113152) that depreciated several nightly APIs.

Furthermore, `proc-macro2` actually _depended_ on one of these APIs.

This was all very promptly addressed and fixed in `proc-macro2` versions `>=1.0.60` (see https://github.com/dtolnay/proc-macro2/pull/391 )

Attempting to consume corrosion with rust nightly toolchain `>=nightly-2023-06-28` will result in an breaking error something like the following:

```console
error[E0635]: unknown feature `proc_macro_span_shrink`
  --> /home/gutenfries/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.52/src/lib.rs:92:30
   |
92 |     feature(proc_macro_span, proc_macro_span_shrink)
   |                              ^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0635`.
error: could not compile `proc-macro2` (lib) due to previous error
error: failed to compile `corrosion-generator v0.1.0 (/home/gutenfries/Desktop/mille/build/linux/x64/release/_deps/corrosion-src/generator)`, intermediate artifacts can be found at `/home/gutenfries/Desktop/mille/build/linux/x64/release/_deps/corrosion-src/generator/target`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
CMake Error at /home/gutenfries/Desktop/mille/build/linux/x64/release/_deps/corrosion-src/cmake/Corrosion.cmake:594 (message):
  Building CMake Generator for Corrosion - failed
Call Stack (most recent call first):
  /home/gutenfries/Desktop/mille/build/linux/x64/release/_deps/corrosion-src/CMakeLists.txt:73 (include)
```

## Additional Context

My CMake call looks like this: _(Note that this is NOT the issue)_

```cmake
# We include Corrosion inline here, but ideally in a project with many
# dependencies we would need to install Corrosion on the system. See
# instructions on https://github.com/AndrewGaspar/corrosion#cmake-install Once
# done, uncomment this line: find_package(Corrosion REQUIRED)

include(FetchContent)

FetchContent_Declare(
  Corrosion
  GIT_REPOSITORY https://github.com/AndrewGaspar/corrosion.git
  GIT_TAG origin/master # Optionally specify a version tag or branch here
)

FetchContent_MakeAvailable(Corrosion)

corrosion_import_crate(MANIFEST_PATH ../native/Cargo.toml)

# Flutter-specific

set(CRATE_NAME "native")

target_link_libraries(${BINARY_NAME} PRIVATE ${CRATE_NAME})

list(APPEND PLUGIN_BUNDLED_LIBRARIES $<TARGET_FILE:${CRATE_NAME}-shared>)
```

## Environment

| tool | version | 
| --- | --- |
| OS | Like Ubuntu 20.04 |
| CMake | 3.16.3 |
| clang++ | 10.0.0-4ubuntu1 |

## Expected Behavior

Not to throw a compile error :smile:

## Steps To Reproduce

Consume Corrorsion in any environment using rust toolchain `nightly-2023-06-28` or higher

## Proposed fix(es)

the issue has  been resolved within `proc-macro2`, and versions `>=1.0.60` appropriately address this.

simply bumping this ASAP should be adequate to address this.

I will open a PR in a minute with the changes.

**Comment by jschwe** [comment]
Will this affect stable starting with Rust 1.72? This is a bit unfortunate, since bumping proc macro2 also means bumping the minimum supported Rust version for the generator.

**Comment by ogoffart** [comment]
> Will this affect stable starting with Rust 1.72? 

No. From my understanding this only affects nightly rust. (proc_maro2 has a build script that detects nightly)

**Comment by gutenfries** [comment]
My apologies on not making it clear, this is ONLY for the _nightly_ toolchain.

> [...] since bumping proc macro2 also means bumping the minimum supported Rust version for the generator.

It is unfortunate, should this move the _(nightly)_ MSRV up, however, from my perspective, the choices are to either let the MSRV go up marginally, or have to enforce a _maximum_ supported nightly version.

> > Will this affect stable starting with Rust 1.72?
> 
> No. From my understanding this only affects nightly rust. (proc_maro2 has a build script that detects nightly)

This is correct.

**Comment by TomzBench** [comment]
I run into this error, but im not sure how to figure out the offending crate

`cargo tree | grep proc-macro2` shows all dependencies seem to be using v1.0.66.

```
│   │   ├── proc-macro2 v1.0.66
│   │   │   └── proc-macro2 v1.0.66 (*)
│   │       ├── proc-macro2 v1.0.66 (*)
│           ├── proc-macro2 v1.0.66 (*)
│               ├── proc-macro2 v1.0.66 (*)
│   │       ├── proc-macro2 v1.0.66 (*)
    │   ├── proc-macro2 v1.0.66 (*)
│               ├── proc-macro2 v1.0.66 (*)
│               │   ├── proc-macro2 v1.0.66 (*)
│       ├── proc-macro2 v1.0.66 (*)
```

But my build error shows 
```error[E0635]: unknown feature `proc_macro_span_shrink`
  --> /home/thomas/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.58/src/lib.rs:92:30
   |
92 |     feature(proc_macro_span, proc_macro_span_shrink)
```

I saw this before and `cargo update` seemed to make the problem go away but that does not seem to work here

**Comment by jschwe** [comment]
> It is unfortunate, should this move the (nightly) MSRV up,

I don't really care about a nightly MSRV, but updating the lockfile means dropping support for anything before stable Rust 1.56, since the newer version of proc macro2 is Rust 2021 edition.
I added a workaround in #430 to avoid this by shipping two lockfiles with corrosion and selecting the correct one based on the Rust version, thus preserving compatibility with old stable versions and fixing the build with nightly Rust.

**Comment by ogoffart** [comment]
Idea: corrosion could detect nightly rust and run cargo update in that case. (Possibly with -p proc_macro2)
But maybe that's overkill.
(Personally I'd be fine raising the msrv to 1.56)

**Comment by jschwe** [comment]
This should be fixed by #430 @gutenfries could you verify if current master works for you?

> ```
> FetchContent_Declare(
>  Corrosion
>  GIT_REPOSITORY https://github.com/AndrewGaspar/corrosion.git
>  GIT_TAG origin/master # Optionally specify a version tag or branch here
> )
> ```

FYI, you should probably change master here to the `stable/v0.4.` branch or the `v0.4` tag. 
As announced in the release notes for v0.4, this the last corrosion version supporting older CMake versions and support for the versions will be removed from master in the next weeks or months.
